src/lib.rs

```diff
@@ -4014,36 +4014,94 @@ extern "C" fn fn_decapsulate_key(
 
 #[cfg(feature = "pkcs11_3_2")]
 extern "C" fn fn_verify_signature_init(
-    _s_handle: CK_SESSION_HANDLE,
-    _mechptr: *mut CK_MECHANISM,
-    _key_handle: CK_OBJECT_HANDLE,
-    _signature: *mut CK_BYTE,
-    _singature_len: CK_ULONG,
+    s_handle: CK_SESSION_HANDLE,
+    mechptr: *mut CK_MECHANISM,
+    key_handle: CK_OBJECT_HANDLE,
+    psignature: *mut CK_BYTE,
+    psignature_len: CK_ULONG,
 ) -> CK_RV {
-    CKR_FUNCTION_NOT_SUPPORTED
+    let rstate = global_rlock!(STATE);
+    let mut session = res_or_ret!(rstate.get_session_mut(s_handle));
+    check_op_empty_or_fail!(session; Verify; mechptr);
+    let mechanism: &CK_MECHANISM = unsafe { &*mechptr };
+    let slot_id = session.get_slot_id();
+    let mut token = res_or_ret!(rstate.get_token_from_slot_mut(slot_id));
+    let key = res_or_ret!(token.get_object_by_handle(key_handle));
+    ok_or_ret!(check_allowed_mechs(mechanism, &key));
+    let mech = res_or_ret!(token.get_mechanisms().get(mechanism.mechanism));
+    if mech.info().flags & CKF_VERIFY == CKF_VERIFY {
+        let sig_len = cast_or_ret!(usize from psignature_len);
+        let signature: &[u8] =
+            unsafe { std::slice::from_raw_parts(psignature, sig_len) };
+        let operation =
+            res_or_ret!(mech.verify_signature_new(mechanism, &key, signature));
+        session.set_operation::<dyn VerifySignature>(operation, false);
+
+        #[cfg(feature = "fips")]
+        init_fips_approval(session, mechanism.mechanism, CKF_VERIFY, &key);
+
+        CKR_OK
+    } else {
+        CKR_MECHANISM_INVALID
+    }
```

Comment:
nit: Should we do it the other way round, similarly as other functions that have in this branch more than a few statements, to avoid the need to overly indent the rest of the function?

```suggestion
    if mech.info().flags & CKF_VERIFY != CKF_VERIFY {
        return CKR_MECHANISM_INVALID;
    }
    let sig_len = cast_or_ret!(usize from psignature_len);
    let signature: &[u8] =
        unsafe { std::slice::from_raw_parts(psignature, sig_len) };
    let operation =
        res_or_ret!(mech.verify_signature_new(mechanism, &key, signature));
    session.set_operation::<dyn VerifySignature>(operation, false);

    #[cfg(feature = "fips")]
    init_fips_approval(session, mechanism.mechanism, CKF_VERIFY, &key);

    CKR_OK
```